### PR TITLE
Feat/faster json encoding

### PR DESCRIPTION
**Pull request by luma**
## What?

Professional development via lg...

* add benchmarking of the json and plain text output encoding
* replace the encoding/json based marshaling with a custom one that only falls back to encoding/json when it needs to. This is much faster than the standard one
* back the json encoding with a reusable pool of buffers to avoid making extra allocations
* attempt to build json in place as much as possible to avoid making extra allocations
* avoid allocations in the text output path as well
* handle errors in the write output path (these aren't bubbled up yet)

## Results

These are slightly unscientific as I took them on my laptop so power management and other things will have interfered with the results.

``` bash
benchmark                                                        old ns/op     new ns/op     delta
BenchmarkFieldLogging/Marshalling_JSON-8                         32718         6606          -79.81%
BenchmarkFieldLogging/Marshalling_JSON_parallel-8                10322         2533          -75.46%
BenchmarkPlainTextFieldLogging/Marshalling_Fields-8              22254         21902         -1.58%
BenchmarkPlainTextFieldLogging/Marshalling_Fields_parallel-8     6841          6616          -3.29%

benchmark                                                        old allocs     new allocs     delta
BenchmarkFieldLogging/Marshalling_JSON-8                         166            49             -70.48%
BenchmarkFieldLogging/Marshalling_JSON_parallel-8                166            49             -70.48%
BenchmarkPlainTextFieldLogging/Marshalling_Fields-8              163            175            +7.36%
BenchmarkPlainTextFieldLogging/Marshalling_Fields_parallel-8     163            175            +7.36%

benchmark                                                        old bytes     new bytes     delta
BenchmarkFieldLogging/Marshalling_JSON-8                         25459         4907          -80.73%
BenchmarkFieldLogging/Marshalling_JSON_parallel-8                47192         17820         -62.24%
BenchmarkPlainTextFieldLogging/Marshalling_Fields-8              18679         13336         -28.60%
BenchmarkPlainTextFieldLogging/Marshalling_Fields_parallel-8     36307         26217         -27.79%
```

## Future work

* Handle errors during writes by writting them to stderr
* implement non-reflection text encoding (much faster and fewer allocations) based on the json one in this pr
* Only serialise fields provided to extend once as they never change. Currently all fields are serialised every time
* Create a logging type and only periodically syncs/flushes
* Create a logging type that logs only a subset of entries sampled from the whole

### CODE REVIEW

File: output.go [hunk 1]
```diff
@@ -86,15 +128,27 @@ func RemoveOutput(output io.Writer) {
...
+	for _, p := range outputs {
+		delete(hookFnPointers, p)
+	}
+
+	outputs = make(map[io.Writer]uintptr)
+}
+
 // AddHook causes logging activity to invoke the given hook function
-func AddHook(fn func(*Entry), opts ...func(*Options)) {
+func AddHook(fn hookFn, opts ...func(*Options)) {
```
Comment (chrisbnt) [inline output.go:144]: If we switch this to a private interface, is that going to make it hard/impossible for people to add their own hooks?

File: util.go [hunk 2]
```diff
@@ -50,17 +51,21 @@ func ExtractTrailingFields(
...
 func RenderMessage(args ...interface{}) string {
-	message := ""
-	for _, a := range args {
-		if message != "" {
-			message += " "
+	var message bytes.Buffer
+
+	for i, a := range args {
+		if i > 0 {
+			// message.WriteString(" ")
```
Comment (chrisbnt) [inline util.go]: I think we can remove this, `message.writeByte(' ')` clearly does this job.

**Review by chrisbnt — COMMENTED** [review]
1 question re: hook functions, otherwise LGTM

File: util.go (hunk 2 above)
Comment (luma) [reply util.go]: Removed

File: output.go (hunk 1 above)
Comment (luma) [reply output.go:144]: It shouldn't, as long as their function has the same signature as hookFn

File: util.go (hunk 2 above)
Comment (luma) [reply util.go]: Done

**Pull request merged**